### ISSUE TITLE
Enable full rendering in Chrome 41

### ISSUE THREAD

**Issue by adrake33** [AugurProject/augur#854]
[ryanb EDIT: hijacking top comment: see [updated requirements in a later comment](https://github.com/AugurProject/augur/issues/854#issuecomment-462538583)]

This error is occurring on the main page and preventing rendering in Chrome 41:

![Screen_Shot_2019-01-14_at_1.26.44_PM.png](https://images.zenhubusercontent.com/5bff29dc5cb43a45a18add9f/a9c3af44-9b46-43f9-b456-bc323d64151b)

**Comment by adrake33** [comment]
@joeykrug I talked to @justinbarry about this. He said he fixed one console error, but fixing this specific one will not be quick. He said this is the result of bad packages upstream, which are shipping ES6 code. We can either fix this by transpiling all of our dependencies or by submitting patches to the packages that aren't releasing right (preferably the latter).

I know you said you wanted to prioritize this above the trading page, but Justin is currently assigned 2 chart-related tasks for the trading page, and it would probably be fastest if he did them, since he has worked on the charts previously. Would it be ok if we did this task after he's finished with these trading page tasks?

**Comment by joeykrug** [comment]
Yep that sounds good

**Comment by pgebheim** [comment]
I thought we already uses babel to transpile all the deps. They definitely get bundled...  does this happen on try.augur.net or just dev.augur.net?

**Comment by joeykrug** [comment]
Happens everywhere

**Comment by ryanberckmans** [comment]
**Q: What's stopping Augur UI from working in Chrome 41?**
1. Augur UI loads the external script `https://instant.0xproject.com/instant.js`, which is in es6, and es6 is unsupported by Chrome 41
2. Chrome 41 doesn't support `grid-*` CSS properties, which Augur UI uses for main layout
3. Chrome 41 doesn't support `font-stretch` property, but we use this only in one place (`src/modules/auth/components/common/error-container.styles.less`) and this is unlikely to be an SEO blocker

**Q: How can we fix `instant.js` being in es6 instead of es5?**
Justin committed a [fix](https://github.com/0xProject/0x-monorepo/commit/a8c3b4126e6b62f423a71dba7ba55a85e7e2bde7) to 0x 25 days ago. I tested Justin's fix in augur ui and confirmed the parse errors go away. **This fix isn't deployed yet** (`https://instant.0xproject.com/instant.js` is out of date with `master`); fix will be automatically picked up by augur UI when it's deployed. **I've asked in 0x's #questions to deploy latest instant.js and will own this follow-up**. Another option I don't recommend (because it's easy to wait until the fix is deployed) is to explore including instant.js into our bundle; not sure why they recommend loading it from their CDN, maybe for browser origin purposes.

**Q: How can we fix `grid-*` in Chrome 41?**
afaict, our options to get Augur UI rendering properly in Chrome 41 are to stop using `grid-*` properties, ie. re-implement our main grid layout using another technology; or to get [css-grid-polyfill](https://github.com/FremyCompany/css-grid-polyfill ) working.

I tried loading AUI in Chrome 41 with css-grid-polyfill, but the rendering was still broken. However there was a javascript error `Uncaught Error: No suitable conversion from unit '' to unit 'px'` that's potentially fixable, so I would recommend exploring this polyfill (which could be conditionally loaded for googlebot) as a solution of first resort.

**Q: What if we don't fix these issues?**
We gotta fix instant.js parse error or the site doesn't work at all. Good thing this one is an easy fix.

Googlebot may partially index Augur UI even if we don't fix use of `grid-*` in Chrome 41. The DOM appears well-formed in Chrome 41; googlebot may parse the markets from dom. But the rendered page is a mess; the markets don't display at all.

**Next steps**
1. Ryan will own follow-up with 0x to get instant.js with es5 fix deployed
2. recommend we **reassign this ticket to a CSS expert** to determine how to move forward on lack of grid-*` support in Chrome 41. We might also start using Google Search Console to determine how effectively AUI is indexed in Chrome 41 even with `grid-*`
3. Ryan could also attempt to solve javascript errors with `css-grid-polyfill`, but this is probably best done by a CSS expert

**Methodology**
* browserstack to compare rendering on dev/try.augur.net with my local `master` augur
* caniuse.com to determine what was unsupported in Chrome 41 https://caniuse.com/#compare=chrome+41,chrome+71
* https://www.browseemall.com/Compatibility/ValidateCSS to verify the CSS properties we use that are unsupported by Chrome 41. But, I only checked our `src`; there may be unsupported CSS properties in `node_modules`. I didn't attempt to run this on the built CSS because it's embedded in the js bundle

**Comment by justinbarry** [comment]
Awesome write up and good catch on the css stuff that wasn’t on my radar!

Did you click around in the app a bit. Not all code paths load immediately. @joeykrug  was finding further issues deeper in the app.

**Comment by joeykrug** [comment]
> But the rendered page is a mess; the markets don't display at all.

Yeah, needs to render fully w/o errors

Could try using the polyfill and see what happens, the average user isnt using chrome 41 anyways so shouldnt be effected

**Comment by ryanberckmans** [comment]
> Did you click around in the app a bit. Not all code paths load immediately. ... finding further issues deeper in the app.

@justinbarry I wasn't able to click around the app much in Chrome 41. But, because the investigation focused on a global search of js/css differences in Chrome 41, I think we can reasonably expect the app to fully work after fixing the identified issues. Also I don't think there's other external libs besides instead.js/airswap, and airswap is already es6. There may be weirder stuff like libraries dynamically loading css that's unsupported in Chrome 41. EDIT: also we're not currently transpiling all of augur ui's `node_modules` to es5, which may be a source of further issues.

**Comment by ryanberckmans** [comment]
I patched `css-grid-polyfill` to get it running in chrome 41, but the layout is still screwed up.

0x team hasn't responded yet about deploying latest `instant.js`. Justin got me into the private #Augur channel on their Discord. I'll keep trying.

I think we need to explore deprecating use of css `grid-*` if we want Augur UI to render in chrome 41.

**Comment by justinbarry** [comment]
There are only a few uses of this css property. @phoebemirman @JohnDanz can we replace it with something else?

**Comment by ryanberckmans** [comment]
That'd be great!

**Comment by ryanberckmans** [comment]
## **TL;DR**

Asking for attention of augur-ui expert to deprecate use of css `grid-*` and refactor dataflow so googlebot can see homepage at OnLoad event.

Talked to Paul: will unassign myself from this now and an augur UI expert can proceed when we're ready. Future person: happy to clarify info below or help with any of the SEO/chrome41 bits.

## What we want Googlebot to see

Joey wants Googlebot to see homepage with data loaded; like this:

![image](https://user-images.githubusercontent.com/299324/52601772-97629a00-2e25-11e9-86f8-0cea23266243.png)

## Chrome 41/Googlebot tiers of compatibility for augur UI

I've tried to come up with a system of "tiers" for chrome 41/googlebot support, and associated action items for augur UI to achieve each tier.

  - [X] homepage data loads properly in chrome 41
      - [X] no javascript errors
      - [X] markets fetch and render into DOM
  - [ ] homepage renders properly in chrome 41
      - [ ] replace css `grid-*` with something that works in chrome 41
  - [ ] UI works (deeper than homepage) in chrome 41
      - [ ] needs verification, after UI renders in chrome 41; currently impractical to click around, due to broken layout
  - [ ] googlebot can crawl homepage DOM with no market data loaded
      - [ ] ensure homepage always rendered at OnLoad event[1]
  - [ ] googlebot can screenshot homepage with no market data loaded
      - [ ] ensure disclaimer modal doesn't occlude homepage at time of OnLoad event, eg. disable disclaimer reliably by checking for googlebot using `window.navigator.userAgent`
  - [ ] googlebot can crawl homepage DOM _with_ market data loaded
      - [ ] ensure market list reliably loaded+displayed at OnLoad event
  - [ ] googlebot can screenshot homepage with market data loaded
      * **This is the goal, Joey wants googlebot to see this**
      * no specific steps needed, just need to get previous milestones done

### [1] Problem: googlebot sees loading splash screen, and not homepage, at OnLoad event 

![image](https://user-images.githubusercontent.com/299324/52600664-f8886e80-2e21-11e9-8c74-9048aa59e7f5.png)
⬆️ This image shows googlebot crawling the loading splash screen.


![image](https://user-images.githubusercontent.com/299324/52600730-4309eb00-2e22-11e9-8975-75ced663933e.png)
⬆️ This image shows OnLoad event occurring during splash screen with local chrome devtools


![image](https://user-images.githubusercontent.com/299324/52600747-4ef5ad00-2e22-11e9-8de4-810c56c7bed0.png)
⬆️ This image shows the splash screen completely replacing entire DOM, so that the homepage UI DOM is gone during splash screen, not not simply hidden.

The problem here is that googlebot snapshots the page layout and DOM at OnLoad event, but augur UI typically has a loading splash screen displayed at OnLoad and homepage UI DOM is gone during splash screen.

The render progression is

1. homepage UI displayed with no data
2. white augur logo loading page screen displayed
     * **problem:** OnLoad event typically occurs in here
3. homepage UI displayed with no data
4. homepage UI displayed with loaded market data

## Resources

* Googlebot url to test fetch: `https://dev2.augur.net/index.html?ethereum_node_http=https%3A%2F%2Frinkeby.augur.net%2Fethereum-http&augur_node=wss%3A%2F%2Fdev.augur.net%2Faugur-node#!/markets`
* [Google Search Console](https://www.google.com/webmasters/tools/googlebot-fetch-details?hl=en&authuser=2&siteUrl=https://dev2.augur.net/&path=index.html?ethereum_node_http%3Dhttps%253a%252f%252feth-mainnet.alchemyapi.io%252fjsonrpc%252f7sE1TzCIRIQA3NJPD5wg7YRiVjhxuWAE%26augur_node%3Dwss%253a%252f%252faugur-node.augur.casino%23!/markets&timestamp=1549923334558), ask Pete for access

**Comment by justinbarry** [comment]
It looks like the site is indexed.
![image](https://user-images.githubusercontent.com/328965/54548459-3585da80-4965-11e9-8c28-d1dab0790587.png)

**Comment by pgebheim** [comment]
It's not really indexed though as it doesn't go into any of the markets.

It looks like this is trying to do that be loading the old-school ?_escape_fragment urls, but it shouldn't need to do that anymore.

According to the docs google attempts to crawl and #! urls, so perhaps we need to update to that scheme instead of just #/blah ?

**Comment by pgebheim** [comment]
I'm closing THIS ticket and we're gonna make another one about crawling market pages specifically.
